core/trino-main/src/main/java/io/trino/sql/planner/optimizations/UnaliasSymbolReferences.java

```diff
@@ -996,16 +1041,27 @@ public PlanAndMappings visitJoin(JoinNode node, UnaliasContext context)
             Optional<Symbol> newRightHashSymbol = node.getRightHashSymbol().map(mapper::map);
 
             // rewrite dynamic filters
-            Set<Symbol> added = new HashSet<>();
+            Map<Symbol, DynamicFilterId> canonicalDynamicFilters = new HashMap<>();
             ImmutableMap.Builder<DynamicFilterId, Symbol> filtersBuilder = ImmutableMap.builder();
+            ImmutableMap.Builder<DynamicFilterId, DynamicFilterId> dynamicFilterIdMapBuilder = ImmutableMap.builder();
             for (Map.Entry<DynamicFilterId, Symbol> entry : node.getDynamicFilters().entrySet()) {
                 Symbol canonical = mapper.map(entry.getValue());
-                if (added.add(canonical)) {
+                DynamicFilterId canonicalDynamicFilterId = canonicalDynamicFilters.putIfAbsent(canonical, entry.getKey());
+                if (canonicalDynamicFilterId == null) {
                     filtersBuilder.put(entry.getKey(), canonical);
                 }
+                else {
+                    dynamicFilterIdMapBuilder.put(entry.getKey(), canonicalDynamicFilterId);
+                }
             }
             Map<DynamicFilterId, Symbol> newDynamicFilters = filtersBuilder.build();
 
+            ImmutableMap<DynamicFilterId, DynamicFilterId> dynamicFilterIdMap = dynamicFilterIdMapBuilder.build();
+            if (!dynamicFilterIdMap.isEmpty()) {
+                // update dynamic filter ids on the probe side
+                rewrittenLeft = node.getLeft().accept(this, context.withDynamicFilterIdMap(dynamicFilterIdMap));
```

Comment:
this is exponential (since you have already visited left side). You need to do another, separate pass